File: backend/src/services/branch.service.js
const pool = require('../config/database');

class BranchService {
  async getAll(tenantId, { status, search, page = 1, limit = 20 }) {
    const offset = (page - 1) * limit;
    const conditions = ['b.tenant_id = $1'];
    const params = [tenantId];
    let idx = 2;

    if (search) {
      conditions.push(`(b.name ILIKE $${idx} OR b.code ILIKE $${idx} OR b.address ILIKE $${idx})`);
      params.push(`%${search}%`); idx++;
    }
    if (status) {
      conditions.push(`b.status = $${idx}`);
      params.push(status); idx++;
    }

    const where = conditions.join(' AND ');
    const countResult = await pool.query(`SELECT COUNT(*) FROM branches b WHERE ${where}`, params);

    const result = await pool.query(
      `SELECT b.*,
              COUNT(DISTINCT r.id) as room_count,
              COALESCE(SUM(r.capacity), 0) as total_capacity
       FROM branches b
       LEFT JOIN rooms r ON r.branch_id = b.id
       WHERE ${where}
       GROUP BY b.id
       ORDER BY b.name
       LIMIT $${idx} OFFSET $${idx + 1}`,
      [...params, limit, offset]
    );

    return { total: parseInt(countResult.rows[0].count), page, limit, branches: result.rows };
  }

  async getById(tenantId, branchId) {
    const result = await pool.query(
      `SELECT b.* FROM branches b WHERE b.id = $1 AND b.tenant_id = $2`,
      [branchId, tenantId]
    );
    if (!result.rows.length) return null;

    const rooms = await pool.query(
      'SELECT * FROM rooms WHERE branch_id = $1 ORDER BY name',
      [branchId]
    );

    return { ...result.rows[0], rooms: rooms.rows };
  }

  async create(tenantId, data) {
    const { name, code, address, phone, status = 'active' } = data;
    const result = await pool.query(
      'INSERT INTO branches (tenant_id, name, code, address, phone, status) VALUES ($1,$2,$3,$4,$5,$6) RETURNING *',
      [tenantId, name, code || null, address || null, phone || null, status]
    );
    return result.rows[0];
  }

  async update(tenantId, branchId, data) {
    const { name, code, address, phone, status } = data;
    const result = await pool.query(
      `UPDATE branches SET name=$1, code=$2, address=$3, phone=$4, status=$5, updated_at=NOW()
       WHERE id=$6 AND tenant_id=$7 RETURNING *`,
      [name, code || null, address || null, phone || null, status, branchId, tenantId]
    );
    return result.rows[0];
  }

  async delete(tenantId, branchId) {
    await pool.query('DELETE FROM branches WHERE id=$1 AND tenant_id=$2', [branchId, tenantId]);
  }

  async addRoom(tenantId, branchId, data) {
    const { name, code, capacity, status = 'active' } = data;
    const result = await pool.query(
      'INSERT INTO rooms (tenant_id, branch_id, name, code, capacity, status) VALUES ($1,$2,$3,$4,$5,$6) RETURNING *',
      [tenantId, branchId, name, code || null, capacity || 0, status]
    );
    return result.rows[0];
  }

  async updateRoom(tenantId, roomId, data) {
    const { name, code, capacity, status } = data;
    const result = await pool.query(
      `UPDATE rooms SET name=$1, code=$2, capacity=$3, status=$4, updated_at=NOW()
       WHERE id=$5 AND tenant_id=$6 RETURNING *`,
      [name, code || null, capacity || 0, status, roomId, tenantId]
    );
    return result.rows[0];
  }

  async deleteRoom(tenantId, roomId) {
    await pool.query('DELETE FROM rooms WHERE id=$1 AND tenant_id=$2', [roomId, tenantId]);
  }
}

module.exports = new BranchService();

File: backend/Services/BranchService.cs
using System.Text.Json;
using Dapper;
using LanguageCenter.Api.Infrastructure;

namespace LanguageCenter.Api.Services;

public sealed class BranchService
{
    private readonly DbService _db;
    public BranchService(DbService db) => _db = db;

    public async Task<object> GetAllAsync(int tenantId, IQueryCollection query)
    {
        var p = new DynamicParameters(); p.Add("tenantId", tenantId);
        var conditions = new List<string> { "b.tenant_id=@tenantId" };
        if (!string.IsNullOrWhiteSpace(query["search"])) { conditions.Add("(b.name ILIKE @search OR b.address ILIKE @search OR b.phone ILIKE @search)"); p.Add("search", $"%{query["search"]}%"); }
        var rows = await _db.QueryAsync($"""
            SELECT b.*,
                   COALESCE(json_agg(json_build_object('id', r.id, 'name', r.name, 'code', r.code, 'capacity', r.capacity, 'status', r.status, 'is_active', (r.status='active')) ORDER BY r.name) FILTER (WHERE r.id IS NOT NULL), '[]') AS rooms
            FROM branches b
            LEFT JOIN rooms r ON r.branch_id=b.id AND r.tenant_id=b.tenant_id
            WHERE {string.Join(" AND ", conditions)}
            GROUP BY b.id
            ORDER BY b.name
            """, p);
        return rows;
    }

    public async Task<Dictionary<string, object?>?> GetByIdAsync(int tenantId, int id)
    {
        var branch = await _db.QuerySingleAsync("SELECT * FROM branches WHERE id=@id AND tenant_id=@tenantId", new { tenantId, id });
        if (branch is null) return null;
        branch["rooms"] = await _db.QueryAsync("SELECT * FROM rooms WHERE branch_id=@id AND tenant_id=@tenantId ORDER BY name", new { tenantId, id });
        return branch;
    }

    public async Task<Dictionary<string, object?>?> CreateAsync(int tenantId, JsonElement body)
    {
        return await _db.QuerySingleAsync(
            """INSERT INTO branches (tenant_id, name, code, address, phone, status) VALUES (@tenantId,@name,@code,@address,@phone,COALESCE(@status,'active')) RETURNING *""",
            new { tenantId, name = body.Str("name"), code = body.Str("code"), address = body.Str("address"), phone = body.Str("phone"), status = body.Str("status") ?? (body.Bool("isActive", "is_active") == false ? "inactive" : "active") });
    }

    public async Task<Dictionary<string, object?>?> UpdateAsync(int tenantId, int id, JsonElement body)
    {
        var current = await GetByIdAsync(tenantId, id); if (current is null) return null;
        return await _db.QuerySingleAsync(
            """UPDATE branches SET name=@name, code=@code, address=@address, phone=@phone, status=@status, updated_at=NOW() WHERE id=@id AND tenant_id=@tenantId RETURNING *""",
            new { tenantId, id, name = body.Str("name") ?? current["name"], code = body.Str("code") ?? current["code"], address = body.Str("address") ?? current["address"], phone = body.Str("phone") ?? current["phone"], status = body.Str("status") ?? (body.Bool("isActive", "is_active") is bool active ? (active ? "active" : "inactive") : current["status"]) });
    }

    public async Task<bool> DeleteAsync(int tenantId, int id) => await _db.ExecuteAsync("DELETE FROM branches WHERE id=@id AND tenant_id=@tenantId", new { tenantId, id }) > 0;

    public async Task<Dictionary<string, object?>?> AddRoomAsync(int tenantId, int branchId, JsonElement body)
    {
        return await _db.QuerySingleAsync("""INSERT INTO rooms (tenant_id, branch_id, name, code, capacity, status) VALUES (@tenantId,@branchId,@name,@code,COALESCE(@capacity,0),COALESCE(@status,'active')) RETURNING *""", new { tenantId, branchId, name = body.Str("name"), code = body.Str("code"), capacity = body.Int("capacity"), status = body.Str("status") ?? (body.Bool("isActive", "is_active") == false ? "inactive" : "active") });
    }

    public async Task<Dictionary<string, object?>?> UpdateRoomAsync(int tenantId, int roomId, JsonElement body)
    {
        var current = await _db.QuerySingleAsync("SELECT * FROM rooms WHERE id=@roomId AND tenant_id=@tenantId", new { tenantId, roomId }); if (current is null) return null;
        return await _db.QuerySingleAsync("""UPDATE rooms SET name=@name, code=@code, capacity=@capacity, status=@status, updated_at=NOW() WHERE id=@roomId AND tenant_id=@tenantId RETURNING *""", new { tenantId, roomId, name = body.Str("name") ?? current["name"], code = body.Str("code") ?? current["code"], capacity = body.Int("capacity") ?? (current["capacity"] is null ? 0 : Convert.ToInt32(current["capacity"])), status = body.Str("status") ?? (body.Bool("isActive", "is_active") is bool active ? (active ? "active" : "inactive") : current["status"]) });
    }

    public async Task<bool> DeleteRoomAsync(int tenantId, int roomId) => await _db.ExecuteAsync("DELETE FROM rooms WHERE id=@roomId AND tenant_id=@tenantId", new { tenantId, roomId }) > 0;
}
